File: 1 - Two Sum/Python/solution.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        if len(nums) < 2: return False # We need at least two values in the dict.

        record = {} # Our dictionary

        for i in range(len(nums)): # Iterate through each number in nums.
            # If we find a "match" in the dictionary, we have our return value.
            # Return the index of the match and the current index.
            if target - nums[i] in record: # If we find a "match" in the dict.
                # Return the index of the match and the current index, in that order.
                return [record[target - nums[i]], i]
            else: # Otherwise...
                record[nums[i]] = i # Just add the record to our dict and continue.

        return []   # We should never get here, since we can always assume a
                    # solution exists. However, I included a default return
                    # for simplicity's sake.

File: 1 - Two Sum/C++/solution.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        if (nums.size() < 2) return vector<int>(); // We need at least two numbers.

        map<int, int> record; // Our dictionary.

        for (int i = 0; i < nums.size(); i++) { // Iterate through every index of nums.
            if (record.find(target - nums[i]) != record.end()) { // If we find a match.
                // Return the index of the match and the current index, in that order.
                return vector<int>({record[target - nums[i]], i});
            }
            else { // Otherwise...
                record[nums[i]] = i; // Add our key-value pair to record.
            }
        }

        return vector<int>(); // Default return. We shouldn't ever get here.
    }
};
